### PR TITLE
Fix: Override hparams via CLI

### PR DESCRIPTION
**Pull request by Berezin-Leonid**
Allow CLI arguments to override checkpoint hyperparameters in LightningCLI

## What does this PR do?

Currently, when `LightningCLI` loads a configuration from a checkpoint (using `--ckpt_path`), the hyperparameters stored in the checkpoint overwrite any arguments passed via the command line. This behavior prevents users from overriding specific parameters (e.g., changing the learning rate) when resuming training or fine-tuning.

This PR adjusts the configuration merging logic. Now, arguments provided explicitly via the CLI (**including parameters defined in config files via `--config`**) take precedence over the hyperparameters loaded from the checkpoint.

**Example:**
With this change, the following command will correctly use `lr=0.001` instead of the value stored in `epoch=1.ckpt`:
```bash
python main.py fit --ckpt_path=epoch=1.ckpt --model.learning_rate=0.001
```

### Implementation Details

To achieve this, I modified the loading logic so that hyperparameters from the checkpoint are applied as **parser defaults** rather than being merged directly into the configuration object.

This leverages the standard `jsonargparse` priority order:
1. **Command Line Arguments** (Highest priority)
2. **Config Files** (e.g., `--config`)
3. **Defaults** (now includes Checkpoint Hyperparameters)

This ensures that any value explicitly provided by the user will correctly override the value stored in the checkpoint.

Addressing the conversation in #21255 about CLI override priority.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21455.org.readthedocs.build/en/21455/

### CODE REVIEW

File: src/lightning/pytorch/cli.py [hunk 1]
```diff
@@ -576,12 +586,14 @@ def _parse_ckpt_path(self) -> None:
...
-            hparams = {self.config.subcommand: {"model": hparams}}
+            hparams = {"model": hparams}
             try:
-                self.config = self.parser.parse_object(hparams, self.config)
-            except SystemExit:
+                subparser = parser._subcommands_action._name_parser_map[self.config.subcommand]
+                subparser.set_defaults(hparams)
+            except KeyError:
                 sys.stderr.write("Parsing of ckpt_path hyperparameters failed!\n")
                 raise
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py]: This is changing the prior behavior, going from an exit that prints the usage, to an ugly exception with a long stack trace. The behavior should not change. Maybe it would work correctly by doing
```suggestion
            except KeyError as ex:
                sys.stderr.write("Parsing of ckpt_path hyperparameters failed!\n")
                parser.error(str(ex), ex)
```

File: src/lightning/pytorch/cli.py [hunk 2]
```diff
@@ -576,12 +586,14 @@ def _parse_ckpt_path(self) -> None:
...
+            self.parse_arguments(parser, args)
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py:635]: Like this the entire parsing is done twice. The speed is not a problem, since training/predicting is normally significantly slower than the parsing. However, there might be some edge case where the first parsing fails, even though this second one would succeed. But, I am fine with leaving it like this. If there are problematic edge cases, then we can fix them when they happen.

File: tests/tests_pytorch/test_cli.py [hunk 3]
```diff
@@ -556,9 +556,8 @@ def add_arguments_to_parser(self, parser):
     with mock.patch("sys.argv", ["any.py"] + cli_args):
         cli = CkptPathCLI(BoringCkptPathModel, subclass_mode_model=True)
 
-    assert isinstance(cli.model, BoringCkptPathSubclass)
+    assert not isinstance(cli.model, BoringCkptPathSubclass)
     assert cli.model.hidden_dim == 8
-    assert cli.model.extra is True
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:561]: Why did this change? Old behaviors should not change, unless there was bug. Was there a bug? If not, revert back.

File: tests/tests_pytorch/test_cli.py [hunk 4]
```diff
@@ -522,7 +522,7 @@ def add_arguments_to_parser(self, parser):
     assert cli.config_init.predict.model.layer.out_features == 3
 
     err = StringIO()
-    with mock.patch("sys.argv", ["any.py"] + cli_args), redirect_stderr(err), pytest.raises(SystemExit):
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:525]:
```suggestion
    with mock.patch("sys.argv", ["any.py"] + cli_args), redirect_stderr(err), pytest.raises(SystemExit):
```

Revert this back, since this behavior should not change.

File: tests/tests_pytorch/test_cli.py [hunk 5]
```diff
@@ -588,6 +587,155 @@ def __init__(self, submodule1: LightningModule, submodule2: LightningModule, mai
...
+        "--trainer.limit_val_batches=1",
+        "--trainer.default_root_dir=./",
+    ]
+
+    with mock.patch("sys.argv", ["any.py"] + save_args):
+        cli = LightningCLI(DemoModel)
+
+    checkpoint_path = str(next(Path(cli.trainer.default_root_dir).rglob("*.ckpt")))
+
+    # --- Phase 2: Resume with CLI overrides ---
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]:
```suggestion
    # --- Phase 2: Predict with CLI overrides ---
```
I think "Resume" make people think that fit will be run, but it is a prediction.

File: tests/tests_pytorch/test_cli.py [hunk 6]
```diff
@@ -588,6 +587,155 @@ def __init__(self, submodule1: LightningModule, submodule2: LightningModule, mai
...
+    ]
+
+    with mock.patch("sys.argv", ["any.py"] + override_args):
+        new_cli = LightningCLI(DemoModel)
+
+    # --- Phase 3: Assertions ---
+    target_classes = 10
+    assert new_cli.model.num_classes == target_classes, (
+        f"Checkpoint restoration failed! Expected num_classes {target_classes}, got {new_cli.model.num_classes}"
+    )
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py]: `num_classes=10` is the default of the model and not overridden anywhere. Maybe this assert is unnecessary?

File: tests/tests_pytorch/test_cli.py (hunk 3 above)
Comment (Berezin-Leonid) [reply tests/tests_pytorch/test_cli.py:561]: Yes, the old behavior was effectively a bug. It violated the standard priority `Checkpoint < CLI args`.

Previously, explicit flags like `--model=BoringCkptPathModel` were ignored in favor of the checkpoint class. The updated test confirms that the CLI argument now correctly takes precedence.

This also complements PR #21408: while that PR handles *parameter* adaptation, this ensures the *class itself* can be overridden. Together, they enable a clean Training -> Inference workflow with different classes.

File: src/lightning/pytorch/cli.py (hunk 1 above)
Comment (Berezin-Leonid) [reply src/lightning/pytorch/cli.py]: Done

File: tests/tests_pytorch/test_cli.py (hunk 5 above)
Comment (Berezin-Leonid) [reply tests/tests_pytorch/test_cli.py]: Done

File: tests/tests_pytorch/test_cli.py (hunk 4 above)
Comment (Berezin-Leonid) [reply tests/tests_pytorch/test_cli.py:525]: Done

File: tests/tests_pytorch/test_cli.py (hunk 6 above)
Comment (Berezin-Leonid) [reply tests/tests_pytorch/test_cli.py]: Done. Removed the redundant assertion.

**Comment by Berezin-Leonid** [comment]
I've addressed all the feedback
Ready for review!

File: tests/tests_pytorch/test_cli.py (hunk 3 above)
Comment (mauvilsa) [reply tests/tests_pytorch/test_cli.py:561]: One drawback of setting hparams as defaults is that the typical [left-to-right parsing priority of command line arguments](https://jsonargparse.readthedocs.io/en/v4.45.0/#override-order) is not respected. For example, a command like `--model=BoringCkptPathModel --ckpt_path=...` can be confusing, because `--ckpt_path` appears after `--model`, and arguments to the right usually take precedence. Since the checkpoint specifies a particular model, it would be logical for it to override the model argument. It might make sense to require that if `--ckpt_path` is provided, it should be the first argument after the subcommand. This would maintain the convention that arguments on the right have higher priority. However, enforcing `--ckpt_path` as the first argument could be difficult to implement cleanly.

Ignoring the order of command line arguments, the reason for including the `--model=BoringCkptPathModel` argument in this test is that the model is required. Without this argument, parsing fails due to the missing model. In fact, this is a case where the initial parsing fails, but the second parsing (after applying checkpoint hparams defaults) succeeds. Ideally, if a checkpoint is provided, specifying a model argument should not be mandatory, since the checkpoint already contains one. The model argument could be optional and used to override the checkpoint's model if given, but not required. However, implementing this might be tricky. The parser must still mark the model as required; otherwise, the output of `--help` would be misleading. One possible solution is to temporarily patch the subparsers during parsing to make the model argument optional, and then, if the model is not provided, call `parser.error`.

File: src/lightning/pytorch/cli.py [hunk 7]
```diff
@@ -576,12 +586,16 @@ def _parse_ckpt_path(self) -> None:
                     "class_path": hparams.pop("_class_path"),
                     "dict_kwargs": hparams,
                 }
-            hparams = {self.config.subcommand: {"model": hparams}}
+            hparams = {"model": hparams}
             try:
-                self.config = self.parser.parse_object(hparams, self.config)
-            except SystemExit:
+                if parser._subcommands_action is None:
+                    return
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py]: How come you added this? The `if` condition in line 575 is supposed to already return when no subcommands are defined. Maybe the condition is not robust enough, since it would not work when a model has a `subcommand` parameter. Still, if there are no subcommands, why execute the lines from 577 up to here? Maybe better instead to change line 575 to `if parser._subcommands_action is None or not self.config.get("subcommand"):`.

File: tests/tests_pytorch/test_cli.py (hunk 3 above)
Comment (Berezin-Leonid) [reply tests/tests_pytorch/test_cli.py:561]: Got it, now I see the intent behind this check. I’ve updated the logic to follow your suggestion: now the model requirement is temporarily relaxed if --ckpt_path is detected in the arguments.

I also added a test case with a "garbage" checkpoint to make sure the parser still throws an error if the model can't be resolved.

File: src/lightning/pytorch/cli.py (hunk 7 above)
Comment (Berezin-Leonid) [reply src/lightning/pytorch/cli.py]: Agreed, it's cleaner. That check was just to satisfy MyPy, but I've already updated it to your suggested version in the latest commit.
